### PR TITLE
PR: Add check for strings missing translations

### PR DESCRIPTION
**Pull request by goanpeca**
# Description

## Type of change

- [x] New feature (non-breaking change which adds functionality)

# References

# How has this been tested?

- [x] example: all tests pass with my change

## Final checklist:
- [x] My PR is the minimum possible work for the desired functionality
- [x] If I included new strings, I have used `trans.` to make them localizable.
      For more information see our [translations guide](https://napari.org/docs/dev/guides/translations.html).

### CODE REVIEW

**Comment by goanpeca** [comment]
This adds a script to check for strings that might need translation. I used the tokenizer and AST to be able to figure things out however I had to turn all strings that where split into a single long stirng, to be able to run the checks and be sure I updated all strings.

I am still looking into a way of avoiding this :-|

**Comment by Carreau** [comment]
This conflict, 

I fixed it for you but can't push... You can find it on my repo under f11f3c9c5e62325f757cb6206d335de593cc3846
or redo it, it's only a couple of  issues `delayed=True` keyword, and a couple of renamed parameters in strings.

**Comment by goanpeca** [comment]
Thanks @carreau, added you as collaborator of the repo, can you push directly now?

File: tools/check_strings.py
Comment (sofroniewn) [inline tools/check_strings.py]: Can we move these strings to their own file without the code at the bottom, maybe `_check_strings_constants` or something? I think it will be easier to maintain the list of strings and the code are seperate

File: tools/check_strings.py
Comment (sofroniewn) [inline tools/check_strings.py]: can you add a returns

File: tools/check_strings.py
Comment (sofroniewn) [inline tools/check_strings.py]: Can you add a doc string here

File: tools/check_strings.py
Comment (sofroniewn) [inline tools/check_strings.py]: Doc string, also including what get's raised. Am I right in thinking that if string is on the skip list but not found it will raise an issue? I think that will help us keep the list of skipped strings up to date too

**Review by sofroniewn — APPROVED** [review]
Thanks @goanpeca this looks great. The strings themselves look good, but i've made some minor requests around putting the skipped strings in their own file, and if the tests also check if the skipped strings are no longer present (which will be important for maintenance).

Maybe after that change is made @tlambert03 and @jni could have a quick look over this too as it will be adding an additional step for contributors to do before all tests pass.

**Comment by sofroniewn** [comment]
In the interest of coordinating PRs and trying to reduce merge conflicts, I think we should try and get this PR in next!

File: tools/check_strings.py
Comment (goanpeca) [reply tools/check_strings.py]: @sofroniewn nothing gets raised here, we collect them and only do a `sys.exit(1)` which should fail on CI so that we can provide a full list of issues.

Added docs.

File: tools/check_strings.py
Comment (goanpeca) [reply tools/check_strings.py]: Indeed!

File: tools/check_strings.py
Comment (goanpeca) [reply tools/check_strings.py]: Fixed.

File: tools/check_strings.py
Comment (goanpeca) [reply tools/check_strings.py]: Ok will do.

Comment (goanpeca) [reply tools/check_strings.py]: Fixed!

File: tools/check_strings.py
Comment (goanpeca) [reply tools/check_strings.py]: Fixed!

**Comment by goanpeca** [comment]
>... and if the tests also check if the skipped strings are no longer present (which will be important for maintenance).

Working on this part @sofroniewn.

File: napari/_qt/__init__.py
Comment (Carreau) [inline napari/_qt/__init__.py:37]: `QtCore.__version__` need to become a parameter.

File: napari/_qt/widgets/qt_dims.py
Comment (Carreau) [inline napari/_qt/widgets/qt_dims.py]: defered ?

File: napari/_vispy/vispy_scale_bar_visual.py
Comment (Carreau) [inline napari/_vispy/vispy_scale_bar_visual.py:173]: Note to self, reviewed up to here.

File: napari/_vispy/vispy_welcome_visual.py
Comment (Carreau) [inline napari/_vispy/vispy_welcome_visual.py]: Is this file meant to be here ?

Comment (sofroniewn) [reply napari/_vispy/vispy_welcome_visual.py]: I think this is our old welcome visual which isn't in use anymore - we can probably delete

File: napari/components/viewer_model.py
Comment (Carreau) [inline napari/components/viewer_model.py]: why not
```suggestion
                msg = trans._(
                    "Plugin {plugin!r} does not provide sample data named {sample!r}. No plugin samples have been registered.",
                    deferred=True,
                    plugin=plugin,
                    sample=sample,
                    samples=samples,
                )
```
?

And `samples` is not used.

File: napari/components/viewer_model.py
Comment (Carreau) [inline napari/components/viewer_model.py:761]: same question.

File: napari/components/viewer_model.py
Comment (Carreau) [reply napari/components/viewer_model.py]: note: I get why we don't use expression like in f-strings; but I thoughts that formatter like `!r`, `%02d` and similar would work.

**Comment by goanpeca** [comment]
Thanks for the review @carreau, will fix the issues!

File: napari/layers/base/base.py
Comment (Carreau) [inline napari/layers/base/base.py:659]: Note to self. Read down to here.

**Comment by Carreau** [comment]
> Thanks for the review @Carreau, will fix the issues!

I can get a shot at it as well, I'm going to get a reading pass first.

File: napari/layers/shapes/_shapes_models/_polgyon_base.py
Comment (Carreau) [inline napari/layers/shapes/_shapes_models/_polgyon_base.py:72]: number=len(data)

File: napari/layers/tracks/tracks.py
Comment (Carreau) [inline napari/layers/tracks/tracks.py:389]: See question about !r above if there is a resolution.

**Comment by goanpeca** [comment]
@Carreau actually I should be able to add a check for missing vars (or extra ones) inside a given string! Thanks for the checks, you gave me the idea :)

File: napari/layers/tracks/tracks.py
Comment (goanpeca) [reply napari/layers/tracks/tracks.py:389]: Yes you are right, I should have left that inside... I am not sure what I was thinking... maybe I was missing my ☕

File: napari/_qt/__init__.py
Comment (goanpeca) [reply napari/_qt/__init__.py:37]: Fixed!

File: napari/_qt/widgets/qt_dims.py
Comment (goanpeca) [reply napari/_qt/widgets/qt_dims.py]: Fixed!

File: napari/_vispy/vispy_welcome_visual.py
Comment (goanpeca) [reply napari/_vispy/vispy_welcome_visual.py]: Ahhh, I put it back in on the rebase. My bad!

File: napari/components/viewer_model.py
Comment (goanpeca) [reply napari/components/viewer_model.py]: Yes agreed!, will revert all of these!

File: napari/_vispy/vispy_welcome_visual.py
Comment (goanpeca) [reply napari/_vispy/vispy_welcome_visual.py]: Fixed

File: napari/components/viewer_model.py
Comment (goanpeca) [reply napari/components/viewer_model.py]: Fixed

File: napari/layers/shapes/_shapes_models/_polgyon_base.py
Comment (goanpeca) [reply napari/layers/shapes/_shapes_models/_polgyon_base.py:72]: Fixed

File: napari/layers/tracks/tracks.py
Comment (goanpeca) [reply napari/layers/tracks/tracks.py:389]: Fixed, added a check and discovered a few others :)

File: napari/components/viewer_model.py
Comment (goanpeca) [reply napari/components/viewer_model.py:761]: Fixed

**Comment by goanpeca** [comment]
>Thanks @goanpeca this looks great. The strings themselves look good, but i've made some minor requests around putting the skipped strings in their own file, and if the tests also check if the skipped strings are no longer present (which will be important for maintenance).

Implemented the requested changes @sofroniewn !

File: tools/check_strings.py
Comment (sofroniewn) [inline tools/check_strings.py]: @jni can you look through this file

**Review by sofroniewn — APPROVED** [review]
Thanks for those changes @goanpeca this now looks done for me. Maybe @jni can look at `check_strings.py` and then we merged.

One thing that came up with @jni in another convo was if we might not want this check to run on every PR in case the contributor / developer burden becomes too high, helping new comers add the trans, and we just run it nightly. I'd like us to try having it on for every PR for a little and if the burden becomes too great we consider that as a fallback option. Curious what @tlambert03 thinks here too?

Right now the PR has no merge conflicts and so I'd like to get this merged as our very next PR because it touches so many files!!

**Comment by goanpeca** [comment]
>One thing that came up with @jni in another convo was if we might not want this check to run on every PR in case the contributor / developer burden becomes too high, helping new comers add the trans, and we just run it nightly. I'd like us to try having it on for every PR for a little and if the burden becomes too great we consider that as a fallback option. Curious what @tlambert03 thinks here too?

Agreed. 

Maybe we do not fail for now, and make sure things are up to date prior a release?

I am happy to help keep this moving while I can do a small "training" meeting. That can come with the whole language packs release process.

**Comment by sofroniewn** [comment]
> Maybe we do not fail for now, and make sure things are up to date prior a release?

Yes I think that might be more gentle. Maybe we disable the test from running each time now and then i'm sure we can merge this tonight. Turning test on later after some training sounds good

**Comment by goanpeca** [comment]
>Yes I think that might be more gentle. Maybe we disable the test from running each time now and then i'm sure we can merge this tonight. Turning test on later after some training sounds good

Or we an just not fail (`sys.exit(0)`) for now, so the team also can see what the output is as PRs are coming. @sofroniewn ?

**Comment by sofroniewn** [comment]
> Or we an just not fail (sys.exit(0)) for now, so the team also can see what the output is as PRs are coming. @sofroniewn ?

Yeah that seems better - what about emitting warnings?

**Review by tlambert03 — APPROVED** [review]
This is pretty heroic @goanpeca.  Learned a lot here seeing how this is done.  Amazing stuff.

I'm approving, since I can see that this would be a terribly annoying PR to have to keep up-to-date.  I admit to a small amount of apprehension about the impact of all this on contributing to napari... but I can also see that if we're serious about offering a translatable app, this is a necessary step.  So LGTM :)

> we might not want this check to run on every PR in case the contributor / developer burden becomes too high, helping new comers add the trans, and we just run it nightly.

This would be my preference.  It feels like a lot to ask every contributor to have to explicitly skip any non-gui string literal in `strings_list.py`.  I know that means that someone else is stuck with a little extra work from time-to-time, perhaps before release or something... but i feel like we're otherwise going to be explaining this on a _lot_ of PRs.

File: napari/plugins/__init__.py
Comment (Carreau) [inline napari/plugins/__init__.py:313]: Will that work with deferend ? Wouldn't the "Join" reify the messages ?

File: napari/plugins/_tests/test_sample_data.py
Comment (Carreau) [inline napari/plugins/_tests/test_sample_data.py:30]: debug ?

Comment (goanpeca) [reply napari/plugins/_tests/test_sample_data.py:30]: Well it depends... when this was failing having the print provides a way to actually fix it, so that is why I added it :)

File: napari/utils/perf/_compat.py
Comment (Carreau) [inline napari/utils/perf/_compat.py:6]:
```
setup.cfg
38:python_requires = >=3.7
```

So this is likely always true. Also on Python 3+ you don't need the `[:2]`.

Might be for a separate PR.

File: napari/utils/translations.py
Comment (Carreau) [inline napari/utils/translations.py:690]: if not translated can we leave it as a multiline-string ? or maybe you do that to avoid error from your script I guess...

**Review by Carreau — APPROVED** [review]
I think if there are issues we can fix later; but with all the files this touches +1 to get it in soon.

File: napari/plugins/__init__.py
Comment (Carreau) [reply napari/plugins/__init__.py:313]: more specifically is `''.join(...)` a TranslationString or a str ? You may need a custom "join", method.

**Comment by goanpeca** [comment]
I talked with @sofroniewn and @tlambert03 and @carreau, for further ideas. I had a path to move forward

- Update this -> https://github.com/napari/napari/blob/master/docs/developers/release.md

Thanks for the feedback.

**Comment by jni** [comment]
Yeah, I am also pro-release-time-batching of this work. A training meeting helps maintainers, but it will not help everyone coming in to contribute to napari. Having a maintainer run this test, and updating all the untranslated strings accumulated since the last release, is a better model imho.

So, all good with this, but I'd really like this to be moved to its own action that runs nightly or on pushes to master. Having a separate action is also helpful because we can point the release guide to the exact link to check the current missing strings.

Thoughts? @goanpeca do you want to do this pre-merge or in a new PR? Either way is fine by me as long as we're all agreed on the way forward. (ie @sofroniewn @tlambert03 feel free to merge this if it's holding up other merges!)

**Comment by goanpeca** [comment]
I can move to a nightly that works too, however if it will be part of the release process it could run locally as well so I am ok with whatever. I Don't see much value in running it on CI if it becomes a step in the release process :)

**Comment by jni** [comment]
> I Don't see much value in running it on CI if it becomes a step in the release process :)

It improves visibility for people other than the release manager. =)

**Comment by jni** [comment]
e.g. if there needs to be a discussion about a particular translation failure/issue, there can be a link to the failing line.

**Comment by goanpeca** [comment]
>It improves visibility for people other than the release manager. =)

Ok :)

**Comment by sofroniewn** [comment]
That proposal is fine with me. Looks like we're all on the same page here with what we want. Once those last changes come in I will merge this PR. Thanks all for the discussion and review, and @goanpeca for all the hard work here!!

File: napari/plugins/__init__.py
Comment (goanpeca) [reply napari/plugins/__init__.py:313]: Yeah I might need a custom one. I will do that in a separate PR https://github.com/napari/napari/issues/2599

File: napari/utils/perf/_compat.py
Comment (goanpeca) [reply napari/utils/perf/_compat.py:6]: Noted on #2599

File: napari/utils/translations.py
Comment (goanpeca) [reply napari/utils/translations.py:690]: Fixed!

File: .github/workflows/test_translations.yml
Comment (goanpeca) [inline .github/workflows/test_translations.yml:6]: @jni 1 am UTC works :) ?

File: docs/developers/release.md
Comment (goanpeca) [inline docs/developers/release.md:53]: Worth a read @sofroniewn

**Comment by goanpeca** [comment]
@tlambert03 so the tests are now being grabbed by tox, any suggestions on skipping them on the pull request workflow but not skipping them on the scheduled one?

Thanks!

**Comment by tlambert03** [comment]
> any suggestions on skipping them on the pull request workflow but not skipping them on the scheduled one?

couple options.  

- we can add '--ignore tools' to the pytest command in tox
- we could just make the `tox` tests dedicated to napari only by setting `PYTEST_PATH = napari` in the tox `setenv` section, (and then have to run `test_strings.py` manually with pytest in a dedicated step or workflow or something).
- we could add another `env` var to the `Test with tox` portion of `test_pull_requests.yml`, and then add something to the `[gh-actions:env]` portion of `tox.ini` to set a tox factor accordingly? 


first one is probably easiest... just have to remember not to start adding tests to tools that we actually _do_ want run by tox?

File: docs/developers/release.md
Comment (sofroniewn) [reply docs/developers/release.md:53]: Looks good to me, thanks for details!

**Comment by goanpeca** [comment]
I like option 1 thanks!

**Comment by Carreau** [comment]
Is that maybe tings we can refine in separate PRs? I think the core of the PR (updating the strings everywhere), is already good and will trigger a large range of merge conflicts.

**Comment by tlambert03** [comment]
> Is that maybe tings we can refine in separate PRs? I think the core of the PR (updating the strings everywhere), is already good and will trigger a large range of merge conflicts.

also fine by me (but if you want to just throw `--ignore tools` into [these two lines](https://github.com/napari/napari/blob/c02403ede6d5c7fe594049b711b4460da4ad03bb/tox.ini#L89-L90), that's fine too.

**Comment by goanpeca** [comment]
Example of the current output of each of the tests @sofroniewn

Screen Shot 2021-04-22 at 6 38 14 PM
Screen Shot 2021-04-22 at 6 38 23 PM

Screen Shot 2021-04-22 at 6 38 08 PM

We are good to go :)

**Comment by tlambert03** [comment]
lol... your strings 😂 
great job on this @goanpeca!

**Comment by goanpeca** [comment]
>lol... your strings 😂

We have to keep it fun 🙃 

>great job on this @goanpeca!

Thanks for all the help! This was an interesting AST learning experience.

**Comment by sofroniewn** [comment]
All tests have passed! Will merge now. Thanks all

**Comment by jni** [comment]
Whooooooo! 🎉 

I'm really happy with where we have ended up after all the discussion.

**Pull request merged**
